### DISCUSSION TITLE
Financial plan goals creation helper

### DISCUSSION THREAD

**Discussion by nmathey** [MadeInPierre/finalynx#26] [Ideas]
Hi,

Maybe this is what you have in mind behind CoPilot/Simulators features, but here are my thoughts.
With the current "folder logic" of finalynx, next step would be to add a layer for financial goals and by by requesting some data to the user, generating the goal meta data to follow it up and even simulating its progression in future.

Here a two exemple of must common goals:

1. Emergency found - Data needed:
- Current monthly saving power
- Current monthly cost of living (mandatory expenses only)
- Risk score : Some personal situation info in order to assess the size of this found (primary revenue stability, family/children at charge, family as a backup, in case you lose your primary revenue source how fast you can found another one, etc.)
Results: 
- Found goal = Monthly cost of living * Risk score 
- (Simulating) End of goal date  = Found goal / Current monthly saving power

2. Retirement found - Data needed:
-  % of current monthly saving power available you want to put in this goal
- Current monthly cost of life (mandatory expenses + "pleasure" expenses)
- Estimated years left to retire
- Estimate years left to 'end of life'
- Estimated average yearly inflation from today to age of retirement
- Estimated average investments performances
Result: 
- Found goal = ((Monthly cost of living + Inflation rate impact in years left to work) - Calculated monthly retirement revenus ) * years to live while retired
- (Simulating) Ideal saving power to consume to achieve this goal, as of today (without power saving optimization - see below) but taking into consideration estimated average investments performances.

Here is a Gsheet where I tried to reply this logic (sorry in French):
https://docs.google.com/spreadsheets/d/1EaIs450fSmNybXIn1Sl11jPJXNUPn1DSnWv0idDIRhA/edit?usp=sharing

Others goals could be: Kid studies found, mortage rimbursment, short term project, financial freedom, etc.
Basically every goals "consuming" saving power.

The step further would be to optimize the saving power all along the the financial plan life using a mix prioritizing goals that should be terminated firsts and snowball method: as soon as one goal is close, its dedicated power saving is added to the next one in the priority list, and so one.

Does it make sense?
Nick

**Comment by MadeInPierre** [comment]
Hi Nick! 

Thanks for your message. I'd argue that the Discussions space is more informal, no biggie for French content :slightly_smiling_face: We can also continue this conversation in French if you prefer!

It looks like we have pretty close visions. Here are my general planned development steps:
- [X] **Portfolio:** Declare a custom tree and set fixed targets to each line, done.
- [ ] **Open to GitHub (ongoing):** Setup everything for a team-ready repo (linting, CI/CD, tests, docs, ...)
- [ ] **Analyzer:** Show graphs I'd like to see in Finary such as:
    - A per-folder asset allocation view.
    - A fund availability allocation (e.g. lines in a PER are unavailable until an event "retirement" or "buy house" is triggered, lines in a PEA are blocked for 5 years, AVs will have a highly taxed state instead of blocked, ...). This graph would show how much money would be available when.
    - Probably more but I haven't thought much about it yet.
- [ ] **Simulator:** Start with a simple "fixed % per year" model and:
    - Define our evolution bets for each line (e.g. `inflation - 1%` for a fond euro, `8%` for stocks, etc). Some easy improvements could include confidence levels, like setting a worst-case and best-case bet for each line, or defining individual scenarios.
    - Declare `Events` (e.g. spend 10k for a project in 2030, buy a home with a loan in 2035, set the "retirement" event in 2060, ...).
        - Events can trigger `Models` (e.g. when I buy a home with loan, gotta pay back each month for 25 years)
    - Add :sparkles: `Goals` :sparkles: to your simulation (e.g. Emergency fund, Retirement plan, kids' studies, ...) from a collection available in Finalynx:
        - Each objective will select one of several `Models`, available to choose from in Finalynx also. Hence, you will have the choice to set your own percentage for retirement with Model `A`, or let Finalynx find it for you with Model `B`. The models in your GSheets could be a good starting point!
    - `Targets` will probably become the static version of `Goals`, e.g. you can set a fixed emergency fund or enable a `Goal` that calculates it for you (and it may evolve through time! :smile:)
    - A final nice graph will show a timeline of our portfolio's evolution with markers at each event. We might also apply filters from `analyzer` to show some colors (e.g. available/blocked portions)
- [ ] **Copilot:** kinda like a fake-AI-thingy where Finalynx would generate a list of transactions each month (i.e. "how much money to move where") to maximize the simulation's end amount. 
    - E.g. If we have two ETF Worlds (one in a PEA and another in a PER), copilot will optimally distribute our monthly investments wile making sure all events remain feasible.
    - I'm excited for this: Finalynx will naturally optimize both short-term and long-term goals at the same time (e.g. restore safety net first if you had an accident) and let you know each month if your life goals are still feasible.
- [ ] **Enhancements:** Have fun with Monte Carlo, write a full-fledged exact optimization engine for Copilot, extensions for other people's situations, optimize events to find when is the best time to buy your house... the sky's the limit! (and my free time :sweat_smile:)

I still haven't written a detailed "vision" of what I have planned for now. The only drafty content is [this TODO file](https://github.com/MadeInPierre/finalynx/blob/main/doc/TODO.md), but I'll soon migrate it to an [issue-based project view](https://github.com/users/MadeInPierre/projects/4/views/1) (maybe it will complement this message). I'll write a separate post with a cleaner version of this "global plan" soon, it'll probably help us discuss and align our ideas.

Unfortunately, `analyzer`, `simulator` and `copilot` are all empty so it'll be difficult to start contributing right away with a future-proof architecture. Let me know what you think of this plan, and we'll see how we can get things started!

**Reply by nmathey** [reply]
Let's continue it in French and switch back to English if non French-speaking folks come in ;)

Yes ce que j'ai en tête semble complémentaire et nous avons la même vision!
Pour ma part je ne suis pas Dev du tout donc le peu de code que je pourrais faire sera clairement à retravailler/nettoyer/optimiser.
~~Ce que je te propose c'est de quand même publier sous GitHub "mon brouillon actuel" pour te faire une idée de ce que je recherche pour mon cas.~~ Ok fait: https://github.com/nmathey/FiPlan

Pour l'organisation des différents ToDo via "l'issue-based view" comment compte tu intégrer les idées: rajouter un bucket dans le kanban nommé "Whishlist"?

J'en profite pour en ajouter une justement car c'est ma manière de fonctionner :)
Pourquoi ne pas passer directement sur un format JSON pour la construction du portfolio individuel (l'équivalent de ton demo.py)? Ainsi c'est plus human-readible qu'un bout de code Python et nous pourrons ainsi l'ouvrir aux autres techno web (type JS) pour la représentation graphique?

Encore une fois je ne suis pas dev et pas forcément du temps tout le temps mais je suis chaud pour aider :)

**Reply by MadeInPierre** [reply]
Plusieurs sujets :
- **Qualité :** Aucun souci pour la qualité du code, le plus important c'est les modèles et les idées ^^ Si ça ne te dérange pas que je retravaille éventuellement tes contributions, pas de problème de mon côté.
- **Code :** Juste pour être sûr, c'est bien normal que ton code ne tourne pas ([erreur ligne 180](https://github.com/nmathey/FiPlan/blob/99fb36d405e19ac9fbaef22d185b55797ac46c1b/FiPlan.py#L179)) ? En tout cas ça a l'air très sympa, je pense qu'on pourra faire quelque chose de similaire ici ! ^^ avec une déclaration en direct Python pour l'instant (et JSON plus tard). Je regarde plus en détail dans le week-end
- **Organisation :** Je t'invite à créer une issue séparée pour chaque idée qui te vient à l'esprit (une issue = une nouvelle feature/bugfix et version). Je propose de répertorier chaque issue en fonction des [milestones](https://github.com/MadeInPierre/finalynx/milestones?direction=asc&sort=title&state=open) sur le [projet principal](https://github.com/users/MadeInPierre/projects/4/views/1). On pourra ensuite assigner qui se motive à développer quelle issue et mettre le statut qu'il faut pour qu'on soit tous au courant :smile: J'ajouterai l'explication ici #30 
- **JSON :** Très bonne idée, je n'en avais pas l'utilité jusqu'à maintenant mais il serait bien de faire une sorte d'API pour permettre des extensions. Faudra probablement en discuter dans un sujet séparé.

Heureux d'avancer dessus à plusieurs en tout cas :smile:

**Comment by nmathey** [comment]
Parfait dès que j'ai un peu de temps je m'attaquerai à la création des idées et je te dirais sur quoi je me sentira capable d'avancer.
Très clairement tu auras du boulot sur la qualité de code mais cela aura au moins le mérite de "faire le job" :)

PS: oui j'étais en pleine modifications quand j'ai chargé le projet et depuis je n'ai plus eu le temps d'y revenir. Te tiens au jus quand c'est de nouveau fonctionnel.

**Reply by nmathey** [reply]
Mon code est corrigé mais l'appli pas finalisée. Cependant cela peut te donner des idées.
Je viens de poster une première "issue" dis moi si le format est ok.

**Reply by MadeInPierre** [reply]
Mes excuses pour mon délai de réponse ! Je me suis un peu perdu dans les issues/tâches :sweat_smile: 

J'avais fait un tour de ton projet et j'aime bien les idées ! Ca me prend beaucoup plus de temps que prévu d'ajouter des features plus petites mais importantes je trouve pour rendre le projet un peu plus pratique (genre calculer combien investir dans chaque ligne pour suivre nos objectifs fixés).

Je reviens vers tes idées dès que je pense m'approcher des fonctions plus complexes de simulation ! :)

**Reply by nmathey** [reply]
Pas de souci depuis nos premiers échanges je n'ai pas bcp de temps à partager non plus :/
